proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,223 @@
+If DXIL is given, then all of the typed buffer `type_traits` will be applied on each
+typed buffer HLSL resource type. Otherwise, the raw buffer type_traits will be 
+applied to each resource type. If a `type_trait` is not true for the given 
+RET, a corresponding error message will be emitted.
+
+The list of type_traits that will be available for use are described below:
+| type_trait | Description|
+|-|-|
+| `__is_complete_type` | An RET should either be a complete type, or a user defined type that has been completely defined. |
+| `__is_intangible` | An RET should be an arithmetic type, or a bool, or a vector or matrix or UDT containing such types. This is equivalent to validating that the RET is not intangible. |
+| `__builtin_is_homogenous` | A typed buffer RET with the DXIL IR target should never have two different subelement types. |
+
+For the SPIR-V IR target, only `__is_complete_type` and `!__is_intangible` 
+need to be true. When the target IR is DXIL, and the resource is a typed buffer variant,
+`__builtin_is_homogenous` will be used to ensure homogeneity. 
+It will use `BuildFlattenedTypeList` to retrieve a small vector of the subelement types.
+From this subvector, the first element will be compared to all elements in the vector,
+and any mismatches will return false.
+Typed buffer RETs with the DXIL IR target will need have a vector length that is
+at most 4, and the total size in bytes is at most 16. However, type_traits are not
+needed to verify these, since they can be checked directly using template techniques
+and the `sizeof` builtin.
+
+* Examples:
+```
+// targeting DXIL
+struct oneInt {
+	int i;
+};
+struct twoInt {
+   int aa;
+   int ab;
+};
+struct a {
+   oneInt bx;
+   int i;
+};
+struct b;
+struct c {
+  oneInt ca;
+  float1 cb;
+};
+struct d {
+  twoInt x[2];
+  twoInt y[2];
+};
+RWBuffer<double2> r0; // valid - RET fits in 4 32-bit quantities
+RWBuffer<int> r1; // valid
+RWBuffer<float> r2; // valid
+RWBuffer<float4> r3; // valid
+RWBuffer<oneInt> r4; // valid
+RWBuffer<oneInt> r5; // valid - all fields are valid primitive types
+RWBuffer<a> r6; // valid - all leaf types are valid primitive types, and homogenous
+
+RWBuffer<b> r7; // invalid - the RET isn't complete, the definition is missing. 
+// the type_trait that would catch this is `__is_complete_type`
+
+RWBuffer<c> r8; // invalid - struct `oneInt` has int types, and this is not homogenous with the float1 contained in `c`. 
+// the type_trait that would catch this is `__builtin_is_homogenous`
+
+StructuredBuffer<c> r8Structured; // valid
+
+RWBuffer<d> r9; // invalid - the struct f cannot be grouped into 4 32-bit quantities.
+// no type_trait would catch this, but it would be caught by a concept failure, using the sizeof builtin.
+
+StructuredBuffer<d> r9Structured; // valid
+
+RWBuffer<RWBuffer<int> > r10; // invalid - the RET has a handle with unknown size, thus it is an intangible RET.
+// the type trait that would catch this is `!__is_intangible`
+```

```

Comment:
How about:

```
struct EightHalves { half x[8] };  // sizeof(EightHalves) == 16
RWBuffer<EightHalves> b;
```

I think this should be rejected because it ends up with 8 components, but I don't see any trait proposed that handles element count.

Reply:
There's no trait that handles element count or total RET size, both of those constraints will be verified by template magic and the `sizeof` builtin. I've added that example to further clarify.

Reply:
Well, magic was a way for me to summarize this line of code in the spec:
`concept is_valid_line_vector = sizeof(T) <= 16 && vector_type_info<T>::Size <= 4;`
This is specifically the plan to implement the element count check and the total size in bytes check.
We get the templated type, T, and run sizeof on it, making sure it's at most 16. Also, if the given templated type is a vector, the Size variable extracts the vector element count, and we make sure that it's at most 4.
`is_valid_line_vector` will only be checked when instantiating RWBuffer when the RET is detected to be a vector type.

Reply:
In the EightHalves example, EightHalves isn't a vector, it is a struct.

Also, how about:

```
struct EightHalvesAnotherWay {
   half a[4];
   half b[3];
   half c;
};
```

Reply:
My sample does seem to fail with UDTs, I've made a big update.